### PR TITLE
refactor: proxy/preview server

### LINKED ISSUE #205: enhance: server proxy
During `wrangler dev`, we start a proxy server to act as an intermediary to the worker running on the edge. This proxy adds relevant headers, and acts as a stable URL during development, which is pretty good. There are some problems with our approach - 
- We only start the server after we receive a token from the preview service, so if we open the browser tab before that we simply see a generic `ERR_CONNECTION_REFUSED` page in the browser (i.e - the server doesn't 'exist')
- When we make a change to the worker, we drop the connection, meaning if someone refreshes the browser they see the same `ERR_CONNECTION_REFUSED` page again. 
- This makes the iteration process feel slow and broken. We should be able to buffer requests while we start/restart the server, so the user's workflow isn't broken. 

We should fix this before GA. I'll get to it at the end if no one else does, but I'm happy to help anyone who'd like to take a crack at it.

### PR DESCRIPTION
**Pull request by threepointone**
Closes #205.

This PR refactors how we setup the proxy server between the developer and the edge preview service during `wrangler dev`. Of note, we start the server immediately. We also buffers requests/streams and hold on to them when starting/refreshing the token. This means a developer should never see `ERR_CONNECTION_REFUSED` error page, or have an older worker respond after making a change to the code. And when the token does get refreshed, we flush said streams/requests with the newer values, making the iteration process a lot smoother and predictable.

### CODE REVIEW

File: packages/wrangler/src/proxy.ts [hunk 1]
```diff
@@ -101,3 +36,264 @@ const HTTP1_HEADERS = new Set([
...
+      requestResponseBufferRef.current = [];
+    } else {
+      proxy.on("request", handleRequest);
+      cleanupListeners.push(() => proxy.off("request", handleRequest));
+      // flush and replay buffered requests
+      requestResponseBufferRef.current.forEach(({ request, response }) =>
+        handleRequest(request, response)
+      );
+      requestResponseBufferRef.current = [];
+    }
```
Comment (petebacondarwin) [inline packages/wrangler/src/proxy.ts:256]: NIT - this whole block could be simplified to remove some of the duplication - for example:

```ts
const actualHandleRequest = assetPath ?
    createHandleAssetsRequest(assetPath, handleRequest) :
    handleRequest;

proxy.on("request", actualHandleRequest);
cleanupListeners.push(() => proxy.off("request", actualHandleRequest));

// flush and replay buffered requests
requestResponseBufferRef.current.forEach(({ request, response }) =>
  actualHandleRequest(request, response)
);
requestResponseBufferRef.current = [];

// ... somewhere else

function createHandleAssetsRequest(assetPath: string, handleRequest: RequestListener) {
  const handleAsset = serveStatic(assetPath, {
    cacheControl: false,
  });
  return (
    request: IncomingMessage,
    response: ServerResponse
  ) => {
    handleAsset(request, response, () => {
      handleRequest(request, response);
    };
  };
}
```

File: packages/wrangler/src/proxy.ts [hunk 2]
```diff
@@ -101,3 +36,264 @@ const HTTP1_HEADERS = new Set([
...
+    const handleUpgrade = (
+      message: IncomingMessage,
+      socket: WebSocket,
+      body: Buffer
+    ) => {
+      const { headers, url } = message;
+      addCfPreviewTokenHeader(headers, previewToken.value);
+      headers["host"] = previewToken.host;
+      const localWebsocket = new WebSocket(message, socket, body);
+      // TODO(soon): Custom WebSocket protocol is not working?
```
Comment (petebacondarwin) [inline packages/wrangler/src/proxy.ts:268]: Is this where we need to handshake over the "Subprotocol negotiation" of websocket that is being requested?

**Review by petebacondarwin — APPROVED** [review]
OK so I found it quite hard to follow what had changed and what was going on - mainly due to my lack of experience in this part of the code base. But I think it looks like it should work.

File: packages/wrangler/src/proxy.ts [hunk 3]
```diff
@@ -101,3 +36,264 @@ const HTTP1_HEADERS = new Set([
...
+        response: ServerResponse
+      ) => {
+        // store the request and response in a buffer so we can replay it later
+        requestResponseBufferRef.current.push({ request, response });
+      };
+
+      proxy.on("request", bufferRequestResponse);
+      cleanupListeners.push(() => proxy.off("request", bufferRequestResponse));
+      return () => {
+        cleanupListeners.forEach((cleanup) => cleanup());
```
[the remark this answers is not in the document]
Comment (threepointone) [reply packages/wrangler/src/proxy.ts:147]: Thinking this out aloud - Every reference inside the effect is dereferenced after this function is called, so I don't think clearing the array would help any more. Specifically, cleanupListeners itself is dreferenced. If any of the listeners itself is caught hanging on to something, then wiping this array wouldn't help anyway.

File: packages/wrangler/src/proxy.ts (hunk 2 above)
Comment (threepointone) [reply packages/wrangler/src/proxy.ts:268]: I'm actually not sure, this cod eis from the previous impl.

File: packages/wrangler/src/proxy.ts (hunk 1 above)
Comment (threepointone) [reply packages/wrangler/src/proxy.ts:256]: I like this!

File: packages/wrangler/src/proxy.ts [hunk 4]
```diff
@@ -1,97 +1,32 @@
...
-  assetPath?: string | null;
-  onRequest?: (headers: IncomingHttpHeaders) => void;
-  onResponse?: (headers: IncomingHttpHeaders) => void;
-}
+import type { CfPreviewToken } from "./api/preview";
+import { useEffect, useRef } from "react";
 
 /**
- * Creates a HTTP/1 proxy that sends requests over HTTP/2.
+ * `usePreviewServer` is a React hook that creates a local development
```
[the remark this answers is not in the document]
Comment (threepointone) [reply packages/wrangler/src/proxy.ts:17]: It's both - the worker we upload gets exposed from the, and we start up a local development server on our own machine that acts as a proxy to it. In local mode, it proxies to miniflare.
